source/slang/parameter-binding.cpp

```diff
@@ -1838,8 +1868,191 @@ void generateParameterBindings(
+    {
+        auto cbInfo = globalVarLayout->findOrAddResourceInfo(LayoutResourceKind::ConstantBuffer);
+        cbInfo->space = 0;
+        cbInfo->index = globalConstantBufferBinding.index;
+    }
+    programLayout->globalScopeLayout = globalVarLayout;
 }
 
+StructTypeLayout* getGlobalStructLayout(
+    ProgramLayout*  programLayout);
+
+RefPtr<ProgramLayout> specializeProgramLayout(
+    TargetRequest * targetReq,
+    ProgramLayout* programLayout, 
+    Substitutions * typeSubst)
+{
+    RefPtr<ProgramLayout> newProgramLayout;
+    newProgramLayout = new ProgramLayout();
+    newProgramLayout->bindingForHackSampler = programLayout->bindingForHackSampler;
+    newProgramLayout->hackSamplerVar = programLayout->hackSamplerVar;
+    for (auto & entryPoint : programLayout->entryPoints)
+    {
+        RefPtr<EntryPointLayout> newEntryPoint = new EntryPointLayout(*entryPoint);
+        // TODO: for now just copy existing entry point layouts, but we eventually need to
+        // specialize these as well...
+        newProgramLayout->entryPoints.Add(newEntryPoint);
+    }
+
+    List<RefPtr<TypeLayout>> paramTypeLayouts;
+    auto globalStructLayout = getGlobalStructLayout(programLayout);
+    SLANG_ASSERT(globalStructLayout);
+    RefPtr<StructTypeLayout> structLayout = new StructTypeLayout();
+    RefPtr<TypeLayout> globalScopeLayout = structLayout;
+    structLayout->uniformAlignment = globalStructLayout->uniformAlignment;
+    
+    // Try to find rules based on the selected code-generation target
+    auto layoutContext = getInitialLayoutContextForTarget(targetReq);
+
+    // If there was no target, or there are no rules for the target,
+    // then bail out here.
+    if (!layoutContext.rules)
+        return newProgramLayout;
+
+ 
+    // we need to initialize a layout context to mark used registers
+    SharedParameterBindingContext sharedContext;
+    sharedContext.compileRequest = targetReq->compileRequest;
+    sharedContext.defaultLayoutRules = layoutContext.getRulesFamily();
+    sharedContext.programLayout = programLayout;
+
+    // Create a sub-context to collect parameters that get
+    // declared into the global scope
+    ParameterBindingContext context;
+    context.shared = &sharedContext;
+    context.translationUnit = nullptr;
+    context.layoutContext = layoutContext;
+    
+    auto constantBufferRules = context.getRulesFamily()->getConstantBufferRules();
+    structLayout->rules = constantBufferRules;
+
+    UniformLayoutInfo structLayoutInfo;
+    structLayoutInfo.alignment = globalStructLayout->uniformAlignment;
+    structLayoutInfo.size = 0;
+    bool anyUniforms = false;
+    Dictionary<RefPtr<VarLayout>, RefPtr<VarLayout>> varLayoutMapping;
+    for (auto & varLayout : globalStructLayout->fields)
+    {
+        // To recover layout context, we skip generic resources in the first pass
+        // If the var is a generic resource, its resourceInfos will be empty.
+        if (varLayout->resourceInfos.Count() == 0)
```

Comment:
This way of detecting things seems a bit fragile, compared to storing something that definitely says “this depends on a generic.” But this is fine for now if it works.